File: ruby/tut4.rb
#!/usr/bin/env ruby

# XMMS2 - X Music Multiplexer System
# Copyright (C) 2003-2006 XMMS2 Team

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.

# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.

# This file is a part of the XMMS2 client tutorial #4.

# For the purposes of this tutorial, the client will retrieve the current
# playlist and get mediainfo for each entry. If your favorite song is the only
# thing in your playlist, load everything from your favorite artist instead.

require 'xmmsclient'

begin
	xmms = Xmms::Client.new('tutorial4').connect(ENV['XMMS_PATH'])
rescue Xmms::Client::ClientError
	puts 'Failed to connect to XMMS2 daemon.'
	puts 'Please make sure xmms2d is running and using the correct IPC path.'
	exit
end

# Like the last tutorial, there will be a medialib_get_info call, but unlike
# tut3, this will get mediainfo for the list of all the IDs in the current
# playlist.
begin
	# Another nice shortcut. The playlist_list method returns an Array of
	# medialib IDs. The array is zero-indexed like any other, so keep this in
	# mind, especially when displaying playlist positions to a user.
	xmms.playlist.entries.wait.value.each_with_index do |id, index|
		# After running the last tutorial, you saw there is really a lot of
		# mediainfo for just one entry. So let's just print out the URL for
		# each playlist entry here so the screen isn't flooded. Note that we
		# don't need to specify the source for the propdict entry - there's
		# only one URL anyway, set by the server, so masquerading the PropDict,
		# we won't miss anything. By now you've probably noticed that all the
		# PropDict keys are lowercase. This is just to make retrieval of values
		# more uniform. (Having keys like :url and :URL would be confusing,
		# no?)

		# Remember to check for a ID of 0--it's invalid.
		if(id.zero?)
			puts "#{index + 1} Invalid ID!"
			next
		end
		begin
			# Remember, the playlist array is zero-indexed!
			puts "#{index + 1}. #{xmms.medialib_get_info(id).wait.value.to_propdict[:url]}"
			# It's possible that the URL isn't defined by the server, like if
			# you added a non-media file to the playlist. You should keep that
			# in mind when writing a client, but in this case PropDict#[:url]
			# will only return a nil value, which just prints as an empty
			# string.
		rescue Xmms::Result::ValueError
			puts "There was an error retrieving mediainfo for ID: #{id}."
		end
	end
rescue Xmms::Result::ValueError
	puts 'There was an error retrieving mediainfo for a playlist entry.'
end

# This is the end of the synchronous clients. You've learned how to deal with
# XMMS2 both by commanding it and handling the data it returns. The next
# tutorial is nothing new in this regard; you will learn how to code
# asynchronously, which just means you will wait for results a bit differently.

File: c++/tut4.cc
/*  XMMS2 - X Music Multiplexer System
 *  Copyright (C) 2003-2006 XMMS2 Team
 * 
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *                   
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  This file is a part of the XMMS2 client tutorial #4
 *  Let's try some lists and to show the playlist
 */

#include <xmmsclient/xmmsclient++.h>
#include <cstdlib>
#include <iostream>

/*
 * This function is basically the same as tut3.cc.
 * But since we're doing it repeatedly in the playlist getter,
 * we move it to a separate function.
 *
 * print out the artist, title and bitrate for each entry
 * in the playlist.
 */
void
get_mediainfo( const Xmms::Client& client, int id )
{

	Xmms::Dict info = client.medialib.getInfo( id );

	std::cout << "artist = ";
	try {
		std::cout << info["artist"] << std::endl;
	}
	catch( Xmms::no_such_key_error& err ) {
		std::cout << "No artist" << std::endl;
	}

	std::cout << "title = ";
	try {
		std::cout << info["title"] << std::endl;
	}
	catch( Xmms::no_such_key_error& err ) {
		std::cout << "Title" << std::endl;
	}

	std::cout << "bitrate = ";

	try {
		std::cout << info["bitrate"] << std::endl;
	}
	catch( Xmms::no_such_key_error& err ) {
		std::cout << "0" << std::endl;
	}

}

int
main()
{

	// The first part of this program is commented in tut1.cc and tut2.cc
	Xmms::Client client( "tutorial4" );

	try {

		client.connect( std::getenv( "XMMS_PATH" ) );

		/*
		 * So let's look at the lists.
		 *
		 * Lists can only contain one type of values. So you either
		 * have a list of strings or a list of ints.  In this case
		 * we ask for the whole current playlist.  It will return a
		 * list of ints.  Each int is the id number of the entry.
		 *
		 * The playlist has two important numbers: the entry and the position.
		 * Each alteration command (move, remove) works on the position
		 * of the entry rather than the id. This is because you can have
		 * more than one item of the same entry in the playlist.
		 *
		 * First we ask for the playlist.
		 */
		Xmms::List< int > list = client.playlist.listEntries();

		/*
		 * Now iterate the list.
		 *
		 * Xmms::List works about the same way as any standard
		 * container bidirectional const_iterator.  It would be a
		 * minor performance loss if one didn't cache end().  It's
		 * because obtaining it from the C library cannot be
		 * optimized out by the compiler.
		 */
		for( Xmms::List< int >::const_iterator i(list.begin()), i_end(list.end()); i != i_end; ++i ) {

			/* Now we have an ID number. Let's feed it to the function
			 * above (which is the same as we learned in tut3.cc)
			 * and print out some info on the entry.
			 */
			get_mediainfo( client, *i );

			/*
			 * Note that the position of the entry is up to you to keep
			 * track of. I suggest that you keep the playlist in a local
			 * data type that is similar to a linked list. This way you
			 * can easily work with playlist updates.
			 *
			 * More about this later.
			 */

		}

	}
	// Handle errors
	catch( Xmms::connection_error& err ) {
		std::cout << "Connection failed: " << err.what() << std::endl;
		return EXIT_FAILURE;
	}
	catch( Xmms::result_error& err ) {
		std::cout << "Error when asking for the playlist, " << err.what()
		          << std::endl;
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;

}

